flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessor.java

```diff
@@ -23,16 +23,12 @@
 import org.apache.flink.runtime.checkpoint.channel.ChannelStateWriter;
```

Comment:
re
> [hotfix] Fix availability handling in TwoInputStreamOperator

0. Can you explain a bit more what's the actual issue (I don't see what is this commit fixing)?
1. Can you create a bug ticket in the jira about about this and change commit from hotfix to [FLINK-XYZ]?
2. Definitely this commit is missing some test. If you will struggle to reproduce (`StreamTwoInputProcessor` is tested on the task level in `StreamTaskSelectiveReadingTest` and `TwoInputStreamTaskTest`. Note that `TwoInputStreamTaskTestHarness` that is used there is in the process of being deprecated and:
  a) it would be just nice to implement the test based on `StreamTaskMailboxTestHarness`. So far it's mostly used for multiple input tasks tests, but it has been used in source tasks and one input tasks tests as well (for example `StreamTaskTest#testNotifyCheckpointOnClosedOperator`, so it should be easy to use for the two inputs as well.
  b) mailbox version of the test harness might be much much better at reproducing a problem that depends on timing as you can execute it step by step
3. it would be best actually to merge this commit in separate PR to verify the performance impact of this bug fix alone (to narrow down suspects in case of regression detected after merging)

Reply:
I had another look at the issue and now that I understand a bit better the assumptions on `getAvailability` I think the current code is ok % maybe we should change `StreamTwoInputProcessor#L206` to:

```
if (status == InputStatus.MORE_AVAILABLE || (status != InputStatus.END_OF_INPUT && (input.isApproximatelyAvailable() || input.isAvailable()))) {
```
My problem was that I did not use the `AvailabilityProvider.AVAILABLE` at all. My code was something like:

```
class MyInput {
   CompletableFuture<?> availabilityMarker = ...

   void executeSomeAsyncOperation() {
      availabilityMarker = new CompletableFuture();
      executeAsync(this::asyncOperation);
   }

   void asyncOperation() {
     ....
     availabilityMarker.complete(null);
   }

   CompletableFuture<?> getAvailability() {
      return availabilityProvider;
   }
}
```

This won't work because the default implementation of `isApproximatelyAvailable()` is never true. I would have to change:

```
   void asyncOperation() {
     ....
     CompletableFuture<?> tmp = availabilityMarker;
     availabilityMarker = AvailabilityProvider.AVAILABLE;
     tmp.complete(null);
   }
```

This is an implicit, not obvious contract that I was unaware of. The change I suggested for `StreamTwoInputProcessor#L206` would probably make it work for my old version, but I think it is not strictly necessary. On the other hand some additional documentation or a walkthrough example could be helpful here, especially as we are exposing this feature in user facing APIs.

I will remove this commit. Sorry for the confusion.

Reply:
Ok, I see, and I agree that your change would increase usability of this code, protecting against those kind of bugs. However as you proposed it would currently add an extra `volatile` check. I think this could be addressed, by combining this check with anti-starvation check. However on the other hand...

It looks like I have implemented it differently in the `StreamMultipleInputProcessor`, where I do not see such check at all, and the processor completely relies on the `InputStatus`, which all in all I think is easier way of handling this. `isAvailable()` is checked only after the first `processInput` call after being woken up:
```
private int selectNextReadingInputIndex() {
	if (!inputSelectionHandler.isAnyInputAvailable()) {
		fullCheckAndSetAvailable();
	}
	...
}
``` 

Regarding the documentation. `AvailabilityProvider` already has quite a bit of java docs, but maybe you could help updating it to avoid such confusion in the future?